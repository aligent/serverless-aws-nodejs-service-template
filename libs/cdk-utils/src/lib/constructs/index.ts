/* v8 ignore start */
export * from './application-stage';
export * from './config-bucket';
export * from './lambda-function';
export * from './ssm-parameter-group';
export * from './step-function-from-file';
export * from './temporary-data-bucket';
