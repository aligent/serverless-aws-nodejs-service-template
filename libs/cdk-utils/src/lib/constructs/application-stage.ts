import { Aspects, Stage, Tags, type StageProps } from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { FunctionBundleAnalyzerAspect } from '../aspects/function-bundle-analyzer';
import { LambdaFunction } from './lambda-function';

abstract class BaseApplicationStage extends Stage {
    static readonly DEFAULT_STAGES = {
        development: 'dev',
        staging: 'stg',
        production: 'prd',
    } as const;
}

type DefaultStageOptions = typeof BaseApplicationStage.DEFAULT_STAGES;
type DefaultStageId = DefaultStageOptions[keyof DefaultStageOptions];

/**
 * Constructor for stages with a custom id.
 *
 * Use of standard stage ids will throw an error.
 */
export class CustomApplicationStage extends BaseApplicationStage {
    constructor(scope: Construct, stage: string, props?: StageProps) {
        if (Object.values(BaseApplicationStage.DEFAULT_STAGES).includes(stage as DefaultStageId)) {
            throw new Error(
                `Stage ID ${stage} is reserved for default stages, use the appropriate constructor`
            );
        }
        super(scope, stage, props);
    }
}

/**
 * Development ApplicationStage
 *
 * Adds an appropriate stage tag.
 * Adds the FunctionBundleAnalyzerAspect.
 */
export class DevelopmentApplicationStage extends BaseApplicationStage {
    constructor(scope: Construct, props?: StageProps) {
        super(scope, BaseApplicationStage.DEFAULT_STAGES.development, props);

        Tags.of(this).add('STAGE', BaseApplicationStage.DEFAULT_STAGES.development);

        // This aspect enables using the `analyze:function` npm script to
        // display bundle analyses for lambda functions
        Aspects.of(this).add(
            new FunctionBundleAnalyzerAspect({
                customFunctionConstruct: LambdaFunction,
            })
        );
    }
}

/**
 * Staging ApplicationStage
 *
 * Adds an appropriate stage tag.
 */
export class StagingApplicationStage extends BaseApplicationStage {
    constructor(scope: Construct, props?: StageProps) {
        super(scope, BaseApplicationStage.DEFAULT_STAGES.staging, props);

        Tags.of(this).add('STAGE', BaseApplicationStage.DEFAULT_STAGES.staging);
    }
}

/**
 * Production ApplicationStage
 *
 * Adds an appropriate stage tag.
 * Constructs can apply production-specific configuration by checking for this stage
 *
 * @example
 * ```ts
 * const stage = Stage.of(scope);
 *
 * if (stage instanceof ProductionApplicationStage) {
 *     // Apply production-specific configuration
 * }
 * ```
 */
export class ProductionApplicationStage extends BaseApplicationStage {
    constructor(scope: Construct, props?: StageProps) {
        super(scope, BaseApplicationStage.DEFAULT_STAGES.production, props);

        Tags.of(this).add('STAGE', BaseApplicationStage.DEFAULT_STAGES.production);
    }
}
