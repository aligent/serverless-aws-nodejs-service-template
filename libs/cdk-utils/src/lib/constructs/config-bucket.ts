import { Duration, RemovalPolicy } from 'aws-cdk-lib';
import { Bucket, type BucketProps } from 'aws-cdk-lib/aws-s3';
import { Construct } from 'constructs';

/**
 * Properties for the TemporaryDataBucket construct
 */
export type ConfigBucketProps = BucketProps;

/**
 * S3 bucket construct for configuration files with versioning
 *
 * This construct creates an S3 bucket with the following defaults:
 * - Versioning enabled
 * - Auto-delete objects when the bucket is removed
 * - Destroy bucket on stack deletion
 * - Retain 5 noncurrent versions
 * - Expire noncurrent versions after 30 days
 *
 * These defaults can be overridden via props if needed.
 */
export class ConfigBucket extends Bucket {
    constructor(scope: Construct, id: string, props?: ConfigBucketProps) {
        const defaultProps = {
            versioned: true,
            autoDeleteObjects: true,
            removalPolicy: RemovalPolicy.DESTROY,
            lifecycleRules: [
                {
                    id: 'ConfigurationFiles',
                    enabled: true,
                    // Keep the 5 most recent versions indefinitely
                    noncurrentVersionsToRetain: 5,
                    // Delete versions beyond the 5 most recent after 7 days
                    noncurrentVersionExpiration: Duration.days(7),
                },
            ],
        } satisfies BucketProps;

        // Merge defaults with user-provided props
        const finalProps = {
            ...defaultProps,
            ...props,
            // If user provides lifecycle rules, merge them with defaults
            // Apply user provided rules first, then defaults
            lifecycleRules: [...(props?.lifecycleRules || []), ...defaultProps.lifecycleRules],
        };

        super(scope, id, finalProps);
    }
}
