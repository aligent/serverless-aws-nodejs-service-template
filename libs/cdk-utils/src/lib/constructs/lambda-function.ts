import { RemovalPolicy, Stage } from 'aws-cdk-lib';
import {
    LogLevel,
    NodejsFunction,
    OutputFormat,
    type NodejsFunctionProps,
} from 'aws-cdk-lib/aws-lambda-nodejs';
import { LogGroup, RetentionDays } from 'aws-cdk-lib/aws-logs';
import { propertyInjectable } from 'aws-cdk-lib/core/lib/prop-injectable';
import { Construct } from 'constructs';
import { DevelopmentApplicationStage, StagingApplicationStage } from './application-stage';

/**
 * LambdaFunctionProps
 *
 * @see https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_lambda_nodejs.NodejsFunctionProps.html
 * for properties inherited from NodejsFunction
 */
export interface LambdaFunctionProps extends NodejsFunctionProps {
    /**
     * The entry point for the Lambda function.
     * This should be the fully resolved path to the handler file.
     *
     * @example path.join(__dirname, 'cdk/handlers/remap-eta.ts')
     */
    readonly entry: string;
    /**
     * Alias for the Lambda function.
     * If undefined, the function will not be versioned or aliased.
     *
     * @example 'LATEST'
     */
    readonly alias?: string;
}

/**
 * Narrowed type helper for setting global defaults
 *
 * Exclusions:
 * logGroup, entry point - doesn't make sense to set on a global level
 * logRetention - legacy API
 *
 * @see https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_lambda_nodejs.NodejsFunctionProps.html
 * for properties inherited from NodejsFunction
 */
type LambdaFunctionContext = Omit<LambdaFunctionProps, 'entry' | 'logGroup' | 'logRetention'>;

/**
 * LambdaFunction construct
 *
 * Set application-level defaults using the context api
 *
 * @example
 * ```ts
 * const APPLICATION_CONTEXT = {
 *     ...LambdaFunction.defineContext({
 *         runtime: Runtime.NODEJS_20_X,
 *     }),
 * };
 * ```
 *
 * @see https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_lambda_nodejs.NodejsFunctionProps.html
 * for properties inherited from NodejsFunction
 */
@propertyInjectable
export class LambdaFunction extends NodejsFunction {
    readonly PROPERTY_INJECTION_ID = '@aligent.cdk-utils.LambdaFunction';
    static readonly CONTEXT_KEY = '@aligent.cdk-utils.LambdaFunction';

    /**
     * Define context properties for the LambdaFunction construct
     *
     * @param props - The properties for the LambdaFunction construct
     * @returns The context for the LambdaFunction construct
     *
     * @see https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_lambda_nodejs.NodejsFunctionProps.html
     * for properties inherited from NodejsFunction
     */
    static defineContext(props: LambdaFunctionContext) {
        return {
            [this.CONTEXT_KEY]: props,
        };
    }

    constructor(scope: Construct, id: string, props: LambdaFunctionProps) {
        // Get base defaults from context
        const defaults: LambdaFunctionContext =
            scope.node.tryGetContext(LambdaFunction.CONTEXT_KEY) || {};

        // Get the current stage
        const currentStage = Stage.of(scope);
        if (!currentStage) {
            throw new Error('This construct must be used within a CDK Stage');
        }

        // Create a log group for the function with stage-appropriate properties if one hasn't been provided
        const logGroup =
            props.logGroup ||
            new LogGroup(scope, `/aws/lambda/${id}`, logGroupProperties(currentStage));

        // Apply stage-specific bundling optimizations
        const bundling = {
            ...bundlingProperties(currentStage),
            ...defaults.bundling,
            ...props.bundling,
        };

        const environment = {
            ...defaults.environment,
            ...props.environment,
            // Always enable source maps in the environment if they've been set for bundling
            ...(bundling.sourceMap ? { NODE_OPTIONS: '--enable-source-maps' } : {}),
        };

        // Merge all configurations with stage-optimized bundling taking precedence
        // TODO may need to do a proper deep merge here
        const finalProps = {
            ...defaults,
            ...props,
            environment,
            bundling,
            logGroup,
        };

        super(scope, id, finalProps);

        if (props.alias) {
            this.addAlias(props.alias);
        }
    }
}

/**
 * Get stage-specific log group properties
 *
 * Development and staging log groups don't need to last forever
 * and should be cleaned up along with their stack when removed
 *
 * @param stage - The stage to get the log group properties for
 * @returns The log group properties for the stage
 */
function logGroupProperties(stage: Stage) {
    const devConfig = {
        retention: RetentionDays.ONE_WEEK,
        removalPolicy: RemovalPolicy.DESTROY,
    };

    const stagingConfig = {
        retention: RetentionDays.SIX_MONTHS,
        removalPolicy: RemovalPolicy.DESTROY,
    };

    const productionConfig = {
        retention: RetentionDays.TWO_YEARS,
        removalPolicy: RemovalPolicy.RETAIN,
    };

    switch (true) {
        case stage instanceof DevelopmentApplicationStage:
            return devConfig;
        case stage instanceof StagingApplicationStage:
            return stagingConfig;
        default:
            return productionConfig;
    }
}

/**
 * Get stage-specific bundling configuration
 *
 * Production stages use optimized settings for performance
 * Development stages use development-friendly settings
 *
 * @param stage - The stage to get the bundling properties for
 * @returns The bundling properties for the stage
 */
function bundlingProperties(stage: Stage) {
    // Dev configuration optimised for debugging, analysis, and bundling speed
    const devConfig = {
        logLevel: LogLevel.INFO,
        minify: false,
        sourceMap: false,
        metafile: true, // Required for bundle analysis
        format: OutputFormat.ESM,
    } satisfies NodejsFunctionProps['bundling'];

    // Staging/Production configurations optimised for bundle size
    const stagingConfig = {
        keepNames: true,
        logLevel: LogLevel.INFO,
        minify: true,
        sourceMap: true,
        buildArgs: {
            bundle: 'true',
            treeShaking: 'true',
        },
        format: OutputFormat.ESM,
    } satisfies NodejsFunctionProps['bundling'];

    const productionConfig = {
        keepNames: true,
        // TODO is this necessary? It might only apply to ESBuild logs
        logLevel: LogLevel.INFO,
        minify: true,
        // TODO: Source maps are really large, consider disabling for production
        sourceMap: true,
        buildArgs: {
            bundle: 'true',
            treeShaking: 'true',
        },
        format: OutputFormat.ESM,
    } satisfies NodejsFunctionProps['bundling'];

    switch (true) {
        case stage instanceof DevelopmentApplicationStage:
            return devConfig;
        case stage instanceof StagingApplicationStage:
            return stagingConfig;
        default:
            return productionConfig;
    }
}
