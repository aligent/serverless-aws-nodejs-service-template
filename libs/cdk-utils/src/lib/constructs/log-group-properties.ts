import { type Stage, RemovalPolicy } from 'aws-cdk-lib';
import { RetentionDays } from 'aws-cdk-lib/aws-logs';
import { DevelopmentApplicationStage, StagingApplicationStage } from './application-stage';

/**
 * Get stage-specific log group properties
 *
 * Development and staging log groups don't need to last forever
 * and should be cleaned up along with their stack when removed
 *
 * @param stage - The stage to get the log group properties for
 * @returns The log group properties for the stage
 */
export function logGroupProperties(stage: Stage) {
    const devConfig = {
        retention: RetentionDays.ONE_WEEK,
        removalPolicy: RemovalPolicy.DESTROY,
    };

    const stagingConfig = {
        retention: RetentionDays.SIX_MONTHS,
        removalPolicy: RemovalPolicy.DESTROY,
    };

    const productionConfig = {
        retention: RetentionDays.TWO_YEARS,
        removalPolicy: RemovalPolicy.RETAIN,
    };

    switch (true) {
        case stage instanceof DevelopmentApplicationStage:
            return devConfig;
        case stage instanceof StagingApplicationStage:
            return stagingConfig;
        default:
            return productionConfig;
    }
}
