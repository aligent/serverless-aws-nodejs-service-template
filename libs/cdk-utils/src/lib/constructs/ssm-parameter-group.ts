import { Function } from 'aws-cdk-lib/aws-lambda';
import { type IStringParameter } from 'aws-cdk-lib/aws-ssm';
import { Construct } from 'constructs';

/**
 * Abstract class to group SSM parameters together.
 *
 * This class is used to group SSM parameters together and grant permissions to functions.
 *
 * @example
 * ```ts
 * class ClientParameters extends SsmParameterGroup {
 *     public readonly parameters;
 *
 *     constructor(scope: Construct, id = 'ClientParameters') {
 *         super(scope, id);
 *
 *         const CLIENT_USERNAME = StringParameter.fromStringParameterName(
 *             this,
 *             'clientUsername',
 *             `/path/to/username`
 *         );
 *
 *         this.parameters = {
 *             CLIENT_USERNAME,
 *         } as const;
 *     }
 * }
 *
 * const clientParameters = new ClientParameters(this);
 *
 * clientParameters.grantToFunction(someLambdaFunction, 'read');
 */
export abstract class SsmParameterGroup<
    T extends Record<string, IStringParameter> = Record<string, IStringParameter>,
> extends Construct {
    public abstract readonly parameters: T;

    // TODO: Investigate using the parameters/secrets extension for fetching SSM parameters from Lambda
    // https://docs.aws.amazon.com/systems-manager/latest/userguide/ps-integration-lambda-extensions.html

    // TODO: Should this generate a shared role instead?
    // Concerned this might generate more cloudformation than it needs to
    grantToFunction(grantee: Function, permission: 'read' | 'write' | 'readwrite') {
        Object.entries(this.parameters).forEach(([key, param]) => {
            grantee.addEnvironment(key, param.parameterName);
            if (permission === 'write' || permission === 'readwrite') {
                param.grantWrite(grantee);
            }

            if (permission === 'read' || permission === 'readwrite') {
                param.grantRead(grantee);
            }
        });
    }

    grantToFunctions(grantees: Function[], permission: 'read' | 'write' | 'readwrite') {
        grantees.forEach(grantee => {
            this.grantToFunction(grantee, permission);
        });
    }
}
