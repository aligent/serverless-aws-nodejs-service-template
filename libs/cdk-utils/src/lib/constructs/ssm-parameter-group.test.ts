import { App, Stack } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { Code, Function, Runtime } from 'aws-cdk-lib/aws-lambda';
import { StringParameter } from 'aws-cdk-lib/aws-ssm';
import { Construct } from 'constructs';
import { beforeEach, describe, expect, test } from 'vitest';

import { SsmParameterGroup } from './ssm-parameter-group';

class TestParameters extends SsmParameterGroup {
    public readonly parameters;

    constructor(scope: Construct, id: string) {
        super(scope, id);

        const param1 = new StringParameter(this, 'Param1', {
            parameterName: '/test/param1',
            stringValue: 'value1',
        });

        const param2 = new StringParameter(this, 'Param2', {
            parameterName: '/test/param2',
            stringValue: 'value2',
        });

        this.parameters = {
            PARAM_ONE: param1,
            PARAM_TWO: param2,
        } as const;
    }
}

describe('SsmParameterGroup', () => {
    let app: App;
    let stack: Stack;
    let template: Template;
    let testParameters: TestParameters;
    let lambda: Function;
    let lambda2: Function;

    beforeEach(() => {
        app = new App();
        stack = new Stack(app, 'TestStack');
        testParameters = new TestParameters(stack, 'MyTestParameters');
        lambda = new Function(stack, 'MyTestFunction', {
            runtime: Runtime.NODEJS_20_X,
            code: Code.fromInline('foo'),
            handler: 'index.handler',
        });
        lambda2 = new Function(stack, 'MyOtherTestFunction', {
            runtime: Runtime.NODEJS_20_X,
            code: Code.fromInline('foo'),
            handler: 'index.handler',
        });
    });

    const synth = () => Template.fromStack(stack);

    const getFunctionAndPolicySnapshots = () => {
        const functions = template.findResources('AWS::Lambda::Function');
        const policies = template.findResources('AWS::IAM::Policy');
        return { functions, policies };
    };

    test('grantToFunction with "read" permission', () => {
        testParameters.grantToFunction(lambda, 'read');
        template = synth();

        const { functions, policies } = getFunctionAndPolicySnapshots();
        expect(functions).toMatchSnapshot();
        expect(policies).toMatchSnapshot();
    });

    test('grantToFunction with "write" permission', () => {
        testParameters.grantToFunction(lambda, 'write');
        template = synth();

        const { functions, policies } = getFunctionAndPolicySnapshots();
        expect(functions).toMatchSnapshot();
        expect(policies).toMatchSnapshot();
    });

    test('grantToFunction with "readwrite" permission', () => {
        testParameters.grantToFunction(lambda, 'readwrite');
        template = synth();

        const { functions, policies } = getFunctionAndPolicySnapshots();
        expect(functions).toMatchSnapshot();
        expect(policies).toMatchSnapshot();
    });

    test('grantToFunctions grants to multiple functions', () => {
        testParameters.grantToFunctions([lambda, lambda2], 'read');
        template = synth();

        const { functions, policies } = getFunctionAndPolicySnapshots();
        expect(functions).toMatchSnapshot();
        expect(policies).toMatchSnapshot();
    });
});
