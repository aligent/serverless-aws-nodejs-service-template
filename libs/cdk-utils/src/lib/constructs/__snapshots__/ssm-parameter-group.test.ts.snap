// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SsmParameterGroup > grantToFunction with "read" permission 1`] = `
{
  "MyOtherTestFunction7633676A": {
    "DependsOn": [
      "MyOtherTestFunctionServiceRole6B72F272",
    ],
    "Properties": {
      "Code": {
        "ZipFile": "foo",
      },
      "Handler": "index.handler",
      "Role": {
        "Fn::GetAtt": [
          "MyOtherTestFunctionServiceRole6B72F272",
          "Arn",
        ],
      },
      "Runtime": "nodejs20.x",
    },
    "Type": "AWS::Lambda::Function",
  },
  "MyTestFunctionE5665775": {
    "DependsOn": [
      "MyTestFunctionServiceRoleDefaultPolicy11BBEF1A",
      "MyTestFunctionServiceRole1183FD15",
    ],
    "Properties": {
      "Code": {
        "ZipFile": "foo",
      },
      "Environment": {
        "Variables": {
          "PARAM_ONE": {
            "Ref": "MyTestParametersParam137FCD884",
          },
          "PARAM_TWO": {
            "Ref": "MyTestParametersParam24E319C53",
          },
        },
      },
      "Handler": "index.handler",
      "Role": {
        "Fn::GetAtt": [
          "MyTestFunctionServiceRole1183FD15",
          "Arn",
        ],
      },
      "Runtime": "nodejs20.x",
    },
    "Type": "AWS::Lambda::Function",
  },
}
`;

exports[`SsmParameterGroup > grantToFunction with "read" permission 2`] = `
{
  "MyTestFunctionServiceRoleDefaultPolicy11BBEF1A": {
    "Properties": {
      "PolicyDocument": {
        "Statement": [
          {
            "Action": [
              "ssm:DescribeParameters",
              "ssm:GetParameters",
              "ssm:GetParameter",
              "ssm:GetParameterHistory",
            ],
            "Effect": "Allow",
            "Resource": {
              "Fn::Join": [
                "",
                [
                  "arn:",
                  {
                    "Ref": "AWS::Partition",
                  },
                  ":ssm:",
                  {
                    "Ref": "AWS::Region",
                  },
                  ":",
                  {
                    "Ref": "AWS::AccountId",
                  },
                  ":parameter",
                  {
                    "Ref": "MyTestParametersParam137FCD884",
                  },
                ],
              ],
            },
          },
          {
            "Action": [
              "ssm:DescribeParameters",
              "ssm:GetParameters",
              "ssm:GetParameter",
              "ssm:GetParameterHistory",
            ],
            "Effect": "Allow",
            "Resource": {
              "Fn::Join": [
                "",
                [
                  "arn:",
                  {
                    "Ref": "AWS::Partition",
                  },
                  ":ssm:",
                  {
                    "Ref": "AWS::Region",
                  },
                  ":",
                  {
                    "Ref": "AWS::AccountId",
                  },
                  ":parameter",
                  {
                    "Ref": "MyTestParametersParam24E319C53",
                  },
                ],
              ],
            },
          },
        ],
        "Version": "2012-10-17",
      },
      "PolicyName": "MyTestFunctionServiceRoleDefaultPolicy11BBEF1A",
      "Roles": [
        {
          "Ref": "MyTestFunctionServiceRole1183FD15",
        },
      ],
    },
    "Type": "AWS::IAM::Policy",
  },
}
`;

exports[`SsmParameterGroup > grantToFunction with "readwrite" permission 1`] = `
{
  "MyOtherTestFunction7633676A": {
    "DependsOn": [
      "MyOtherTestFunctionServiceRole6B72F272",
    ],
    "Properties": {
      "Code": {
        "ZipFile": "foo",
      },
      "Handler": "index.handler",
      "Role": {
        "Fn::GetAtt": [
          "MyOtherTestFunctionServiceRole6B72F272",
          "Arn",
        ],
      },
      "Runtime": "nodejs20.x",
    },
    "Type": "AWS::Lambda::Function",
  },
  "MyTestFunctionE5665775": {
    "DependsOn": [
      "MyTestFunctionServiceRoleDefaultPolicy11BBEF1A",
      "MyTestFunctionServiceRole1183FD15",
    ],
    "Properties": {
      "Code": {
        "ZipFile": "foo",
      },
      "Environment": {
        "Variables": {
          "PARAM_ONE": {
            "Ref": "MyTestParametersParam137FCD884",
          },
          "PARAM_TWO": {
            "Ref": "MyTestParametersParam24E319C53",
          },
        },
      },
      "Handler": "index.handler",
      "Role": {
        "Fn::GetAtt": [
          "MyTestFunctionServiceRole1183FD15",
          "Arn",
        ],
      },
      "Runtime": "nodejs20.x",
    },
    "Type": "AWS::Lambda::Function",
  },
}
`;

exports[`SsmParameterGroup > grantToFunction with "readwrite" permission 2`] = `
{
  "MyTestFunctionServiceRoleDefaultPolicy11BBEF1A": {
    "Properties": {
      "PolicyDocument": {
        "Statement": [
          {
            "Action": "ssm:PutParameter",
            "Effect": "Allow",
            "Resource": {
              "Fn::Join": [
                "",
                [
                  "arn:",
                  {
                    "Ref": "AWS::Partition",
                  },
                  ":ssm:",
                  {
                    "Ref": "AWS::Region",
                  },
                  ":",
                  {
                    "Ref": "AWS::AccountId",
                  },
                  ":parameter",
                  {
                    "Ref": "MyTestParametersParam137FCD884",
                  },
                ],
              ],
            },
          },
          {
            "Action": [
              "ssm:DescribeParameters",
              "ssm:GetParameters",
              "ssm:GetParameter",
              "ssm:GetParameterHistory",
            ],
            "Effect": "Allow",
            "Resource": {
              "Fn::Join": [
                "",
                [
                  "arn:",
                  {
                    "Ref": "AWS::Partition",
                  },
                  ":ssm:",
                  {
                    "Ref": "AWS::Region",
                  },
                  ":",
                  {
                    "Ref": "AWS::AccountId",
                  },
                  ":parameter",
                  {
                    "Ref": "MyTestParametersParam137FCD884",
                  },
                ],
              ],
            },
          },
          {
            "Action": "ssm:PutParameter",
            "Effect": "Allow",
            "Resource": {
              "Fn::Join": [
                "",
                [
                  "arn:",
                  {
                    "Ref": "AWS::Partition",
                  },
                  ":ssm:",
                  {
                    "Ref": "AWS::Region",
                  },
                  ":",
                  {
                    "Ref": "AWS::AccountId",
                  },
                  ":parameter",
                  {
                    "Ref": "MyTestParametersParam24E319C53",
                  },
                ],
              ],
            },
          },
          {
            "Action": [
              "ssm:DescribeParameters",
              "ssm:GetParameters",
              "ssm:GetParameter",
              "ssm:GetParameterHistory",
            ],
            "Effect": "Allow",
            "Resource": {
              "Fn::Join": [
                "",
                [
                  "arn:",
                  {
                    "Ref": "AWS::Partition",
                  },
                  ":ssm:",
                  {
                    "Ref": "AWS::Region",
                  },
                  ":",
                  {
                    "Ref": "AWS::AccountId",
                  },
                  ":parameter",
                  {
                    "Ref": "MyTestParametersParam24E319C53",
                  },
                ],
              ],
            },
          },
        ],
        "Version": "2012-10-17",
      },
      "PolicyName": "MyTestFunctionServiceRoleDefaultPolicy11BBEF1A",
      "Roles": [
        {
          "Ref": "MyTestFunctionServiceRole1183FD15",
        },
      ],
    },
    "Type": "AWS::IAM::Policy",
  },
}
`;

exports[`SsmParameterGroup > grantToFunction with "write" permission 1`] = `
{
  "MyOtherTestFunction7633676A": {
    "DependsOn": [
      "MyOtherTestFunctionServiceRole6B72F272",
    ],
    "Properties": {
      "Code": {
        "ZipFile": "foo",
      },
      "Handler": "index.handler",
      "Role": {
        "Fn::GetAtt": [
          "MyOtherTestFunctionServiceRole6B72F272",
          "Arn",
        ],
      },
      "Runtime": "nodejs20.x",
    },
    "Type": "AWS::Lambda::Function",
  },
  "MyTestFunctionE5665775": {
    "DependsOn": [
      "MyTestFunctionServiceRoleDefaultPolicy11BBEF1A",
      "MyTestFunctionServiceRole1183FD15",
    ],
    "Properties": {
      "Code": {
        "ZipFile": "foo",
      },
      "Environment": {
        "Variables": {
          "PARAM_ONE": {
            "Ref": "MyTestParametersParam137FCD884",
          },
          "PARAM_TWO": {
            "Ref": "MyTestParametersParam24E319C53",
          },
        },
      },
      "Handler": "index.handler",
      "Role": {
        "Fn::GetAtt": [
          "MyTestFunctionServiceRole1183FD15",
          "Arn",
        ],
      },
      "Runtime": "nodejs20.x",
    },
    "Type": "AWS::Lambda::Function",
  },
}
`;

exports[`SsmParameterGroup > grantToFunction with "write" permission 2`] = `
{
  "MyTestFunctionServiceRoleDefaultPolicy11BBEF1A": {
    "Properties": {
      "PolicyDocument": {
        "Statement": [
          {
            "Action": "ssm:PutParameter",
            "Effect": "Allow",
            "Resource": {
              "Fn::Join": [
                "",
                [
                  "arn:",
                  {
                    "Ref": "AWS::Partition",
                  },
                  ":ssm:",
                  {
                    "Ref": "AWS::Region",
                  },
                  ":",
                  {
                    "Ref": "AWS::AccountId",
                  },
                  ":parameter",
                  {
                    "Ref": "MyTestParametersParam137FCD884",
                  },
                ],
              ],
            },
          },
          {
            "Action": "ssm:PutParameter",
            "Effect": "Allow",
            "Resource": {
              "Fn::Join": [
                "",
                [
                  "arn:",
                  {
                    "Ref": "AWS::Partition",
                  },
                  ":ssm:",
                  {
                    "Ref": "AWS::Region",
                  },
                  ":",
                  {
                    "Ref": "AWS::AccountId",
                  },
                  ":parameter",
                  {
                    "Ref": "MyTestParametersParam24E319C53",
                  },
                ],
              ],
            },
          },
        ],
        "Version": "2012-10-17",
      },
      "PolicyName": "MyTestFunctionServiceRoleDefaultPolicy11BBEF1A",
      "Roles": [
        {
          "Ref": "MyTestFunctionServiceRole1183FD15",
        },
      ],
    },
    "Type": "AWS::IAM::Policy",
  },
}
`;

exports[`SsmParameterGroup > grantToFunctions grants to multiple functions 1`] = `
{
  "MyOtherTestFunction7633676A": {
    "DependsOn": [
      "MyOtherTestFunctionServiceRoleDefaultPolicy646DF3D8",
      "MyOtherTestFunctionServiceRole6B72F272",
    ],
    "Properties": {
      "Code": {
        "ZipFile": "foo",
      },
      "Environment": {
        "Variables": {
          "PARAM_ONE": {
            "Ref": "MyTestParametersParam137FCD884",
          },
          "PARAM_TWO": {
            "Ref": "MyTestParametersParam24E319C53",
          },
        },
      },
      "Handler": "index.handler",
      "Role": {
        "Fn::GetAtt": [
          "MyOtherTestFunctionServiceRole6B72F272",
          "Arn",
        ],
      },
      "Runtime": "nodejs20.x",
    },
    "Type": "AWS::Lambda::Function",
  },
  "MyTestFunctionE5665775": {
    "DependsOn": [
      "MyTestFunctionServiceRoleDefaultPolicy11BBEF1A",
      "MyTestFunctionServiceRole1183FD15",
    ],
    "Properties": {
      "Code": {
        "ZipFile": "foo",
      },
      "Environment": {
        "Variables": {
          "PARAM_ONE": {
            "Ref": "MyTestParametersParam137FCD884",
          },
          "PARAM_TWO": {
            "Ref": "MyTestParametersParam24E319C53",
          },
        },
      },
      "Handler": "index.handler",
      "Role": {
        "Fn::GetAtt": [
          "MyTestFunctionServiceRole1183FD15",
          "Arn",
        ],
      },
      "Runtime": "nodejs20.x",
    },
    "Type": "AWS::Lambda::Function",
  },
}
`;

exports[`SsmParameterGroup > grantToFunctions grants to multiple functions 2`] = `
{
  "MyOtherTestFunctionServiceRoleDefaultPolicy646DF3D8": {
    "Properties": {
      "PolicyDocument": {
        "Statement": [
          {
            "Action": [
              "ssm:DescribeParameters",
              "ssm:GetParameters",
              "ssm:GetParameter",
              "ssm:GetParameterHistory",
            ],
            "Effect": "Allow",
            "Resource": {
              "Fn::Join": [
                "",
                [
                  "arn:",
                  {
                    "Ref": "AWS::Partition",
                  },
                  ":ssm:",
                  {
                    "Ref": "AWS::Region",
                  },
                  ":",
                  {
                    "Ref": "AWS::AccountId",
                  },
                  ":parameter",
                  {
                    "Ref": "MyTestParametersParam137FCD884",
                  },
                ],
              ],
            },
          },
          {
            "Action": [
              "ssm:DescribeParameters",
              "ssm:GetParameters",
              "ssm:GetParameter",
              "ssm:GetParameterHistory",
            ],
            "Effect": "Allow",
            "Resource": {
              "Fn::Join": [
                "",
                [
                  "arn:",
                  {
                    "Ref": "AWS::Partition",
                  },
                  ":ssm:",
                  {
                    "Ref": "AWS::Region",
                  },
                  ":",
                  {
                    "Ref": "AWS::AccountId",
                  },
                  ":parameter",
                  {
                    "Ref": "MyTestParametersParam24E319C53",
                  },
                ],
              ],
            },
          },
        ],
        "Version": "2012-10-17",
      },
      "PolicyName": "MyOtherTestFunctionServiceRoleDefaultPolicy646DF3D8",
      "Roles": [
        {
          "Ref": "MyOtherTestFunctionServiceRole6B72F272",
        },
      ],
    },
    "Type": "AWS::IAM::Policy",
  },
  "MyTestFunctionServiceRoleDefaultPolicy11BBEF1A": {
    "Properties": {
      "PolicyDocument": {
        "Statement": [
          {
            "Action": [
              "ssm:DescribeParameters",
              "ssm:GetParameters",
              "ssm:GetParameter",
              "ssm:GetParameterHistory",
            ],
            "Effect": "Allow",
            "Resource": {
              "Fn::Join": [
                "",
                [
                  "arn:",
                  {
                    "Ref": "AWS::Partition",
                  },
                  ":ssm:",
                  {
                    "Ref": "AWS::Region",
                  },
                  ":",
                  {
                    "Ref": "AWS::AccountId",
                  },
                  ":parameter",
                  {
                    "Ref": "MyTestParametersParam137FCD884",
                  },
                ],
              ],
            },
          },
          {
            "Action": [
              "ssm:DescribeParameters",
              "ssm:GetParameters",
              "ssm:GetParameter",
              "ssm:GetParameterHistory",
            ],
            "Effect": "Allow",
            "Resource": {
              "Fn::Join": [
                "",
                [
                  "arn:",
                  {
                    "Ref": "AWS::Partition",
                  },
                  ":ssm:",
                  {
                    "Ref": "AWS::Region",
                  },
                  ":",
                  {
                    "Ref": "AWS::AccountId",
                  },
                  ":parameter",
                  {
                    "Ref": "MyTestParametersParam24E319C53",
                  },
                ],
              ],
            },
          },
        ],
        "Version": "2012-10-17",
      },
      "PolicyName": "MyTestFunctionServiceRoleDefaultPolicy11BBEF1A",
      "Roles": [
        {
          "Ref": "MyTestFunctionServiceRole1183FD15",
        },
      ],
    },
    "Type": "AWS::IAM::Policy",
  },
}
`;
